smack-extensions/src/main/java/org/jivesoftware/smackx/softwareinfo/element/SoftwareInfoElement.java

```diff
@@ -0,0 +1,219 @@
+
+    public static final String FORMTYPEVALUE = "urn:xmpp:dataforms:softwareinfo";
+    public static final String NAMESPACE = "jabber:x:data";
+    public static final String ELEMENT = "x";
+    private final DataForm form;
+    private final Map<String, FormField> fields;
+
+    private SoftwareInfoElement(Map<String, FormField> formFields) {
+        form = new DataForm(Type.result);
+        form.addFields(formFields.values());
+        this.fields = formFields;
+    }
+
+    @Override
+    public String getNamespace() {
+        return NAMESPACE;
+    }
+
+    @Override
+    public String getElementName() {
+        return ELEMENT;
+    }
+
+    /**
+     * The DataForm returned contains all extension elements as {@link FormField}s.
+     * <br>
+     * Use dataform.getFields() to obtain Software Information.
+     * @return {@link DataForm}
+     */
+    public DataForm getForm() {
+        return form;
+    }
+
+    @Override
+    public int hashCode() {
+        HashCode.Builder builder = HashCode.builder();
+        for (FormField field : fields.values()) {
+            builder.append(field.getValues());
+        }
+        return builder.build();
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        return EqualsUtil.equals(this, obj, (equalsBuilder, otherObj) ->
+                                                                  equalsBuilder.append(fields.get("HEADER"), otherObj.fields.get("HEADER"))
+                                                                               .append(fields.get("icon"), otherObj.fields.get("icon"))
+                                                                               .append(fields.get("os"), otherObj.fields.get("os"))
+                                                                               .append(fields.get("os_version"), otherObj.fields.get("os_version"))
+                                                                               .append(fields.get("software"), otherObj.fields.get("software"))
+                                                                               .append(fields.get("software_version"), otherObj.fields.get("software_version")));
+    }
+
+    @Override
+    public CharSequence toXML(XmlEnvironment xmlEnvironment) {
+        XmlStringBuilder xml = new XmlStringBuilder();
+        xml.append(form);
+        return xml;
+    }
+
+
+    /**
+     * Returns a new instance of {@link Builder}.
+     * @return Builder
+     */
+    public static Builder getBuilder() {
+        return new Builder();
+    }
+
+    /**
+     * This class defines a Builder class for {@link SoftwareInfoElement}
+     * <br>
+     * SoftwareInfoElement.Builder builder = SoftwareInfoElement.getBuilder();
+     * <br>
+     * Values such as header,icon,os,os_version,software and software_version
+     * can be set using their respective setters.
+     */
+    public static final class Builder {
+        Map<String, FormField> fields;
```

Comment:
I don't think that this is the right type for `fields`. Shouldn't it be `DataForm` instead?